### DISCUSSION TITLE
How do I ensure milvus connection at start of the Flask application?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18840] [Q&A and General discussion]
How do I ensure milvus connection at start of the Flask application?
Does connections.connect("default", host=MILVUS_HOST, port=MILVUS_PORT) return something? So I can create something like a singleton?

**Comment by martinenkoEduard** [comment]
Because it seems to me that creating new milvus connection every time, adds overhead.

while True:
    try:
        connections.connect("default", host=MILVUS_HOST, port=MILVUS_PORT)
        collection = Collection("faces")
        print("connected to Milvus")
        break    
    except Exception as e:
        print("Cannot connect to Milvus" + "Error: "  + str(e)  + " stackTrace: " + traceback.format_exc())
        print(f"Cannot connect to Milvus. Trying to connect Again. Sleeping for: {sleep_for_seconds}")
        
        time.sleep(sleep_for_seconds)

Sadly doesn't work.
It says that you must create connection first in some cases.

**Comment by martinenkoEduard** [comment]
werkzeug.exceptions.InternalServerError: 500 Internal Server Error: milvus error inserting Face embedding Error: <ConnectionNotExistException: (code=0, message=should create connect first.)> stackTrace: Traceback (most recent call last):
  File "/app/flask/AdaFace/SQLAlchemy_exampleAdaFace.py", line 415, in detectAndSaveAllFaces
    mr = collection.insert([[face_id],np_arr])
  File "/home/administrator/anaconda3/envs/p37_toch/lib/python3.7/site-packages/pymilvus/orm/collection.py", line 546, in insert
    conn = self._get_connection()
  File "/home/administrator/anaconda3/envs/p37_toch/lib/python3.7/site-packages/pymilvus/orm/collection.py", line 159, in _get_connection
    return connections._fetch_handler(self._using)
  File "/home/administrator/anaconda3/envs/p37_toch/lib/python3.7/site-packages/pymilvus/orm/connections.py", line 336, in _fetch_handler
    raise ConnectionNotExistException(0, ExceptionsMessage.ConnectFirst)
pymilvus.exceptions.ConnectionNotExistException: <ConnectionNotExistException: (code=0, message=should create connect first.)>

**Comment by martinenkoEduard** [comment]
What is the best way to reconnect or ensure shared connection object?

**Reply by XuanYang-cn** [reply]
`connections` is a thread-local singleton.

You could try  `connections.has_connection(alias)` to check if connected

**Reply by martinenkoEduard** [reply]
Does connections.connect(host=_HOST, port=_PORT) always creates a new connection, or does it reuses a existing one?

It's hard to guess how does it work with flask, because it does a lot of under the hood magic to improve concurrency.

Do i ever need to call disconnect explicitly?

What will happen if i don't call disconnect?

**Reply by yhmo** [reply]
> Does connections.connect(host=_HOST, port=_PORT) always creates a new connection, or does it reuses a existing one?
> 
> It's hard to guess how does it work with flask, because it does a lot of under the hood magic to improve concurrency.
> 
> Do i ever need to call disconnect explicitly?
> 
> What will happen if i don't call disconnect?

The connections.connect() use a parameter named "alias" as connection's name to create connection. The connections object maintains a map between name and connections. By default the "alias=default". If you call connections.connect(host=_HOST, port=_PORT), then the connection's name is "default".

To create a new connection with a name:
`connections.connect(host=_HOST, port=_PORT, alias="xxx")`

To use a specified connection, you need to provide a connection name to a Collection object:
`collection = Collection(name=name, using="xxx")`

If you use Collection like this `collection = Collection(name=name)`, it will use the "default" connection.

**Comment by yhmo** [comment]
@martinenkoEduard 
Use this method to retry connection:

```
def create_connection():
    retry = True
    while retry:
        try:
            print(f"Create connection...")
            connections.connect(host=_HOST, port=_PORT)
            retry = False
        except Exception as e:
            print("Cannot connect to Milvus. Error: " + str(e))
            print("Trying to connect Again. Sleeping for: 1 second")
            time.sleep(1)

    print(f"Connect succeed! List connections:")
    print(connections.list_connections())
```
